src/index/txindex.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INDEX_TXINDEX_H
+#define BITCOIN_INDEX_TXINDEX_H
+
+#include <threadinterrupt.h>
+#include <txdb.h>
+#include <uint256.h>
+#include <validationinterface.h>
+
+class CBlockIndex;
+
+/**
+ * TxIndex is used to look up transactions included in the blockchain by hash.
+ * The index is written to a keyspace in the block index database and records
+ * the filesystem location of each transaction by transaction hash.
+ */
+class TxIndex final : public CValidationInterface
+{
+private:
+    const std::unique_ptr<TxIndexDB> m_db;
+
+    /// Whether the index is in sync with the main chain. The flag is flipped
+    /// from false to true once, after which point this starts processing
+    /// ValidationInterface notifications to stay in sync.
+    std::atomic<bool> m_synced;
+
+    /// The last block in the chain that the TxIndex is in sync with.
+    std::atomic<const CBlockIndex*> m_best_block_index;
+
+    std::thread m_thread_sync;
+    CThreadInterrupt m_interrupt;
+
+    /// Initialize internal state from the database and block index.
+    bool Init();
+
+    /// Sync the tx index with the block index starting from the current best
+    /// block. Intended to be run in its own thread, m_thread_sync, and can be
+    /// interrupted with m_interrupt. Once the txindex gets in sync, the
+    /// m_synced flag is set and the BlockConnected ValidationInterface callback
+    /// takes over and the sync thread exits.
+    void ThreadSync();
+
+    /// Write update index entries for a newly connected block.
+    bool WriteBlock(const CBlock& block, const CBlockIndex* pindex);
+
+protected:
+    void BlockConnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex,
+                        const std::vector<CTransactionRef>& txn_conflicted) override;
+
+public:
+    explicit TxIndex(std::unique_ptr<TxIndexDB> db);
```

Comment:
Can you make this forwardable?

```
TxIndex(std::unique_ptr<TxIndexDB> &&db);
```

Or just create the unique_ptr here, see my other comment regarding g_txindex.reset()

Reply:
I find it clearer to pass by value here, as is recommended by [this StackOverflow post](https://stackoverflow.com/a/8114913). Is there a big benefit to changing the parameter to an rvalue-ref?

Reply:
You're right, disregard.